File: ipc-ui/frontend/src/stores/l1-gateways.ts
import { defineStore } from 'pinia'
import { ref, computed, watch } from 'vue'
import { apiService } from '@/services/api'
import { useNetworkStore } from './network'

export interface L1GatewayConfig {
  id: string
  name: string
  address: string
  registry_address: string
  network_id: string
  network_name: string
  chain_id: number
  deployed_at: string
  deployer_address: string
  is_default: boolean
  description?: string
}

export interface L1GatewayConfigFile {
  default_gateway?: string
  gateways: L1GatewayConfig[]
  last_updated: string
}

export const useL1GatewaysStore = defineStore('l1Gateways', () => {
  // State
  const l1Gateways = ref<L1GatewayConfig[]>([])
  const selectedGatewayId = ref<string>('')
  const isLoading = ref(false)
  const error = ref<string | null>(null)

  // Get network store for current network context
  const networkStore = useNetworkStore()

  // Computed
  const selectedGateway = computed(() => {
    return l1Gateways.value.find(gateway => gateway.id === selectedGatewayId.value)
  })

  const availableGateways = computed(() => {
    // Filter gateways for current network
    const currentNetwork = networkStore.selectedNetwork
    if (!currentNetwork) {
      console.log('[L1GatewaysStore] No current network selected, showing all gateways')
      return l1Gateways.value
    }

    // Map network ID to expected subnet path using chainId
    const expectedRootPath = currentNetwork.chainId ? `/r${currentNetwork.chainId}` : null

    const filtered = l1Gateways.value.filter(gateway => {
      // Check if gateway's network_id matches the expected root path for current network
      if (expectedRootPath && gateway.network_id === expectedRootPath) {
        return true
      }

      // Fallback: legacy hardcoded checks for backwards compatibility
      if (currentNetwork.id === 'calibration' && gateway.network_id === '/r314159') {
        return true
      }

      if (currentNetwork.id === 'local-anvil' && gateway.network_id === '/r31337') {
        return true
      }

      return false
    })

    console.log(`[L1GatewaysStore] Filtered ${filtered.length} gateways for network ${currentNetwork.id} (${expectedRootPath})`)
    return filtered
  })

  const defaultGateway = computed(() => {
    return availableGateways.value.find(gateway => gateway.is_default) || availableGateways.value[0]
  })

  // Actions
  const loadL1Gateways = async () => {
    isLoading.value = true
    error.value = null

    try {
      console.log('[L1GatewaysStore] Loading L1 gateways...')

      // Load L1 gateways configuration file
      const response = await apiService.getL1GatewayConfig()

      console.log('[L1GatewaysStore] API response:', response.data)

      // The API returns data in ApiResponse format: { success, data: { gateways, default_gateway, ... } }
      if (response.data?.data?.gateways) {
        l1Gateways.value = response.data.data.gateways
        console.log('[L1GatewaysStore] Loaded gateways:', l1Gateways.value)

        // Set selected gateway from config or default
        if (response.data.data.default_gateway) {
          selectedGatewayId.value = response.data.data.default_gateway
          console.log('[L1GatewaysStore] Set selected gateway from config:', selectedGatewayId.value)
        } else if (defaultGateway.value) {
          selectedGatewayId.value = defaultGateway.value.id
          console.log('[L1GatewaysStore] Set selected gateway to default:', selectedGatewayId.value)
        }
      } else {
        console.warn('[L1GatewaysStore] No gateways found in response')
      }
    } catch (err) {
      console.error('Failed to load L1 gateways:', err)
      error.value = err instanceof Error ? err.message : 'Failed to load L1 gateways'
    } finally {
      isLoading.value = false
    }
  }

  const selectGateway = async (gatewayId: string) => {
    if (selectedGatewayId.value === gatewayId) return

    const gateway = l1Gateways.value.find(g => g.id === gatewayId)
    if (!gateway) {
      throw new Error(`Gateway with ID ${gatewayId} not found`)
    }

    selectedGatewayId.value = gatewayId

    try {
      // Update config.toml with new gateway selection
      await apiService.updateL1GatewaySelection(gatewayId)
      console.log(`Selected L1 gateway: ${gateway.name} (${gateway.address})`)
    } catch (err) {
      console.error('Failed to update gateway selection:', err)
      error.value = err instanceof Error ? err.message : 'Failed to update gateway selection'
      throw err
    }
  }

  const addL1Gateway = async (gateway: Omit<L1GatewayConfig, 'id' | 'deployed_at'>) => {
    try {
      const response = await apiService.addL1Gateway(gateway)

      if (response.data?.gateway) {
        l1Gateways.value.push(response.data.gateway)

        // If this is the first gateway for this network, make it default
        if (availableGateways.value.length === 1) {
          await selectGateway(response.data.gateway.id)
        }
      }

      return response.data?.gateway
    } catch (err) {
      console.error('Failed to add L1 gateway:', err)
      error.value = err instanceof Error ? err.message : 'Failed to add L1 gateway'
      throw err
    }
  }

  const removeL1Gateway = async (gatewayId: string) => {
    try {
      await apiService.removeL1Gateway(gatewayId)

      const index = l1Gateways.value.findIndex(g => g.id === gatewayId)
      if (index >= 0) {
        l1Gateways.value.splice(index, 1)

        // If we removed the selected gateway, select another one
        if (selectedGatewayId.value === gatewayId && defaultGateway.value) {
          await selectGateway(defaultGateway.value.id)
        }
      }
    } catch (err) {
      console.error('Failed to remove L1 gateway:', err)
      error.value = err instanceof Error ? err.message : 'Failed to remove L1 gateway'
      throw err
    }
  }

  const updateGatewayAsDefault = async (gatewayId: string) => {
    try {
      // Remove default from all gateways
      l1Gateways.value.forEach(g => g.is_default = false)

      // Set new default
      const gateway = l1Gateways.value.find(g => g.id === gatewayId)
      if (gateway) {
        gateway.is_default = true
        await selectGateway(gatewayId)
      }
    } catch (err) {
      console.error('Failed to update default gateway:', err)
      throw err
    }
  }

  // Watch network changes to load appropriate gateways
  watch(() => networkStore.selectedNetworkId, async (newNetworkId) => {
    if (newNetworkId) {
      await loadL1Gateways()
    }
  })

  return {
    // State
    l1Gateways: computed(() => l1Gateways.value),
    selectedGatewayId: computed(() => selectedGatewayId.value),
    selectedGateway,
    availableGateways,
    defaultGateway,
    isLoading: computed(() => isLoading.value),
    error: computed(() => error.value),

    // Actions
    loadL1Gateways,
    selectGateway,
    addL1Gateway,
    removeL1Gateway,
    updateGatewayAsDefault
  }
})

File: ipc/cli/src/commands/ui/api/l1_gateways.rs
// Copyright 2022-2024 Protocol Labs
// SPDX-License-Identifier: MIT
//! L1 Gateway configuration API endpoints

use super::super::services::gateway_service::GatewayService;
use super::super::AppState;
use super::types::ApiResponse;
use crate::GlobalArguments;
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::convert::Infallible;
use warp::{Filter, Reply};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct L1GatewayConfig {
    pub id: String,
    pub name: String,
    pub address: String,
    pub registry_address: String,
    pub network_id: String,
    pub network_name: String,
    pub chain_id: u64,
    pub deployed_at: String,
    pub deployer_address: String,
    pub is_default: bool,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct L1GatewayConfigFile {
    pub default_gateway: Option<String>,
    pub gateways: Vec<L1GatewayConfig>,
    pub last_updated: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateGatewaySelectionRequest {
    pub gateway_id: String,
}

impl Default for L1GatewayConfigFile {
    fn default() -> Self {
        L1GatewayConfigFile {
            default_gateway: None,
            gateways: Vec::new(),
            last_updated: chrono::Utc::now().to_rfc3339(),
        }
    }
}

/// Get L1 gateway configuration
pub async fn get_l1_gateway_config(state: AppState) -> Result<impl Reply, warp::Rejection> {
    log::info!("Getting L1 gateway configuration from discovered gateways");

    // Get discovered gateways and convert them to L1Gateway format
    let global = crate::GlobalArguments {
        config_path: Some(state.config_path.clone()),
        _network: fvm_shared::address::Network::Testnet,
        __network: None,
    };

    let gateway_service = crate::commands::ui::services::GatewayService::new(global);

    match gateway_service.discover_gateways(None).await {
        Ok(discovered_gateways) => {
            log::info!(
                "Discovered {} gateways for L1 configuration",
                discovered_gateways.len()
            );

            // Convert discovered gateways to L1GatewayConfig format
            let l1_gateways: Vec<L1GatewayConfig> = discovered_gateways
                .iter()
                .filter(|gw| {
                    // Only include L1 gateways (root network gateways)
                    gw.parent_network.starts_with("/r")
                        && gw.parent_network.matches('/').count() == 1 // e.g., "/r31337"
                })
                .enumerate()
                .map(|(index, gw)| {
                    L1GatewayConfig {
                        id: gw.id.clone(),
                        name: gw
                            .name
                            .clone()
                            .unwrap_or_else(|| format!("Gateway {}", index + 1)),
                        address: gw.address.clone(),
                        registry_address: gw.registry_address.clone(),
                        network_id: gw.parent_network.clone(),
                        network_name: format!(
                            "Local Anvil {}",
                            gw.parent_network.replace("/r", "")
                        ),
                        chain_id: 31337, // Default for local anvil, could be parsed from network_id
                        deployed_at: gw.deployed_at.to_rfc3339(),
                        deployer_address: "unknown".to_string(), // Not available in GatewayInfo
                        is_default: index == 0,                  // Make first gateway default
                        description: Some(format!(
                            "Gateway contract serving network {}",
                            gw.parent_network
                        )),
                    }
                })
                .collect();

            let config = L1GatewayConfigFile {
                default_gateway: l1_gateways.first().map(|gw| gw.id.clone()),
                gateways: l1_gateways,
                last_updated: chrono::Utc::now().to_rfc3339(),
            };

            log::info!(
                "Returning {} L1 gateways in configuration",
                config.gateways.len()
            );

            let response = ApiResponse {
                success: true,
                data: Some(serde_json::to_value(config).unwrap()),
                error: None,
            };

            Ok(warp::reply::json(&response))
        }
        Err(e) => {
            log::error!("Failed to discover gateways for L1 configuration: {}", e);

            // Return empty config on error
            let config = L1GatewayConfigFile::default();
            let response = ApiResponse {
                success: true,
                data: Some(serde_json::to_value(config).unwrap()),
                error: None,
            };

            Ok(warp::reply::json(&response))
        }
    }
}

/// Update L1 gateway selection
pub async fn update_l1_gateway_selection(
    request: UpdateGatewaySelectionRequest,
    _state: AppState,
) -> Result<impl Reply, warp::Rejection> {
    log::info!("Updating L1 gateway selection to: {}", request.gateway_id);

    // For now, just acknowledge the update
    // In a real implementation, this would update the config file
    let response = ApiResponse {
        success: true,
        data: Some(serde_json::json!({
            "selected_gateway": request.gateway_id,
            "updated_at": chrono::Utc::now().to_rfc3339()
        })),
        error: None,
    };

    Ok(warp::reply::json(&response))
}

/// Add a new L1 gateway
pub async fn add_l1_gateway(
    gateway: L1GatewayConfig,
    _state: AppState,
) -> Result<impl Reply, warp::Rejection> {
    log::info!(
        "Adding new L1 gateway: {} ({})",
        gateway.name,
        gateway.address
    );

    // For now, just acknowledge the addition
    // In a real implementation, this would add to the config file
    let response = ApiResponse {
        success: true,
        data: Some(serde_json::to_value(&gateway).unwrap()),
        error: None,
    };

    Ok(warp::reply::json(&response))
}

/// Remove an L1 gateway
pub async fn remove_l1_gateway(
    gateway_id: String,
    _state: AppState,
) -> Result<impl Reply, warp::Rejection> {
    log::info!("Removing L1 gateway: {}", gateway_id);

    // For now, just acknowledge the removal
    // In a real implementation, this would remove from the config file
    let response = ApiResponse {
        success: true,
        data: Some(serde_json::json!({
            "removed_gateway_id": gateway_id,
            "removed_at": chrono::Utc::now().to_rfc3339()
        })),
        error: None,
    };

    Ok(warp::reply::json(&response))
}

/// Helper to pass state to handlers
fn with_state(state: AppState) -> impl Filter<Extract = (AppState,), Error = Infallible> + Clone {
    warp::any().map(move || state.clone())
}

/// Create L1 gateway API routes
pub fn l1_gateway_routes(
    state: AppState,
) -> impl Filter<Extract = impl Reply, Error = warp::Rejection> + Clone {
    let get_config = warp::path!("l1-gateways" / "config")
        .and(warp::get())
        .and(with_state(state.clone()))
        .and_then(handle_get_l1_gateway_config);

    let update_selection = warp::path!("l1-gateways" / "selection")
        .and(warp::put())
        .and(warp::body::json())
        .and(with_state(state.clone()))
        .and_then(handle_update_l1_gateway_selection);

    let add_gateway = warp::path("l1-gateways")
        .and(warp::post())
        .and(warp::body::json())
        .and(with_state(state.clone()))
        .and_then(handle_add_l1_gateway);

    let remove_gateway = warp::path!("l1-gateways" / String)
        .and(warp::delete())
        .and(with_state(state.clone()))
        .and_then(handle_remove_l1_gateway);

    get_config
        .or(update_selection)
        .or(add_gateway)
        .or(remove_gateway)
}

/// Handle get L1 gateway configuration request
async fn handle_get_l1_gateway_config(state: AppState) -> Result<impl Reply, warp::Rejection> {
    get_l1_gateway_config(state).await
}

/// Handle update L1 gateway selection request
async fn handle_update_l1_gateway_selection(
    request: UpdateGatewaySelectionRequest,
    state: AppState,
) -> Result<impl Reply, warp::Rejection> {
    update_l1_gateway_selection(request, state).await
}

/// Handle add L1 gateway request
async fn handle_add_l1_gateway(
    gateway: L1GatewayConfig,
    state: AppState,
) -> Result<impl Reply, warp::Rejection> {
    add_l1_gateway(gateway, state).await
}

/// Handle remove L1 gateway request
async fn handle_remove_l1_gateway(
    gateway_id: String,
    state: AppState,
) -> Result<impl Reply, warp::Rejection> {
    remove_l1_gateway(gateway_id, state).await
}
